include/IECore/LRUCache.inl

```diff
@@ -38,117 +38,605 @@
+			// to `get( someOtherKey )`, and this inner call has
+			// then entered `limitCost()`.
+			typename List::iterator it = list.begin();
+			while( it != list.end() && it->hasHandle )
+			{
+				++it;
+			}
+
+			if( it == list.end() )
+			{
+				return false;
+			}
+
+			const Item &item = *it;
+
+			key = item.key;
+			cacheEntry = item.cacheEntry;
+
+			list.erase( it );
+
+			return true;
+		}
+
+		typename LRUCache::Cost currentCost;
+
+	private :
+
+		MapAndList m_mapAndList;
+
+};
+
+// Uses a binned map to allow concurrent map operations, and
+// uses a second-chance algorithm to avoid the serial operations
+// associated with managing an LRU list.
+template<typename LRUCache>
+class Parallel
+{
+
+	public :
+
+		typedef typename LRUCache::CacheEntry CacheEntry;
+		typedef typename LRUCache::KeyType Key;
+		typedef tbb::atomic<typename LRUCache::Cost> AtomicCost;
+
+		struct Item
+		{
+			Item() : recentlyUsed() {}
+			Item( const Key &key ) : key( key ), recentlyUsed() {}
+			Item( const Item &other ) : key( other.key ), cacheEntry( other.cacheEntry ), recentlyUsed() {}
+			Key key;
+			mutable CacheEntry cacheEntry;
+			// Mutex to protect cacheEntry.
+			typedef tbb::spin_rw_mutex Mutex;
+			mutable Mutex mutex;
+			// Flag used in second-chance algorithm.
+			mutable tbb::atomic<bool> recentlyUsed;
+		};
+
+		// We would love to use one of TBB's concurrent containers as
+		// our map, but we need the ability to insert, erase and iterate
+		// concurrently. The concurrent_unordered_map doesn't provide
+		// concurrent erase, and the concurrent_hash_map doesn't provide
+		// concurrent iteration. Instead we choose a non-threadsafe
+		// container, but split our storage into multiple bins with a
+		// container in each bin. This way concurrent operations do not
+		// contend on a lock unless they happen to target the same bin.
+		typedef boost::multi_index::multi_index_container<
+			Item,
+			boost::multi_index::indexed_by<
+				// Equivalent to std::unordered_map, using Item::key
+				// as the key. This actually has a couple of benefits
+				// over std::unordered_map :
+				//
+				// - Insertion does not invalidate existing iterators.
+				//   This allows us to store m_popIterator.
+				// - Lookup can be performed using types other than the
+				//   key. This provides the possibility of creating a
+				//   prehashed key prior to taking a Bin lock, although
+				//   this is not implemented here yet.
+				boost::multi_index::hashed_unique<
+					boost::multi_index::member<Item, Key, &Item::key>
+				>
+			>
+		> Map;
+
+		typedef typename Map::iterator MapIterator;
+
+		struct Bin
+		{
+			Bin() {}
+			Bin( const Bin &other ) : map( other.map ) {}
+			Bin &operator = ( const Bin &other ) { map = other.map; return *this; }
+			Map map;
+			typedef tbb::spin_rw_mutex Mutex;
+			Mutex mutex;
+		};
+
+		typedef std::vector<Bin> Bins;
+
+		Parallel()
+		{
+			m_bins.resize( tbb::tbb_thread::hardware_concurrency() );
```

Comment:
It definitely doesn't account for things like our `-threads` argument. It _might_ account for a limited CPU count from something like cgroups, but I don't know. I don't think it's something to worry about too much though...